pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go

```diff
@@ -0,0 +1,349 @@
+		}
+
+		return []reconcile.Request{
+			{NamespacedName: types.NamespacedName{Namespace: mabc.Namespace, Name: mabc.Name}},
+		}
+	})
+}
+
+type reconciler struct {
+	logger        *logrus.Entry
+	client        ctrlruntimeclient.Client
+	buildClient   *buildv1client.BuildV1Client
+	architectures []string
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	logger := r.logger.WithField("request", req.String())
+	err := r.reconcile(ctx, req, logger)
+	if err != nil {
+		logger.WithError(err).Error("Reconciliation failed")
+	} else {
+		logger.Info("Finished reconciliation")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+func (r *reconciler) reconcile(ctx context.Context, req reconcile.Request, logger *logrus.Entry) error {
+	logger = logger.WithField("multiarchbuildconfig_name", req.Name)
+	logger.Info("Starting reconciliation")
+
+	mabc := &v1.MultiArchBuildConfig{}
+	if err := r.client.Get(ctx, req.NamespacedName, mabc); err != nil {
+		return fmt.Errorf("failed to get the MultiArchBuildConfig: %w", err)
+	}
+
+	if mabc.Status.State == v1.SuccessState || mabc.Status.State == v1.FailureState {
+		return nil
+	}
+
+	if err := r.handleMultiArchBuildConfig(ctx, mabc); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *reconciler) handleMultiArchBuildConfig(ctx context.Context, mabc *v1.MultiArchBuildConfig) error {
+	builds := make(map[string]*buildv1.Build)
+	targetImageRef := fmt.Sprintf("%s/%s", mabc.Spec.BuildSpec.CommonSpec.Output.To.Namespace, mabc.Spec.BuildSpec.CommonSpec.Output.To.Name)
+	r.logger.WithField("target_image", targetImageRef).Info("Start handling builds to generate image...")
+
+	for _, arch := range r.architectures {
+		commonSpec := mabc.Spec.BuildSpec.CommonSpec.DeepCopy()
+		commonSpec.NodeSelector = map[string]string{nodeArchitectureLabel: arch}
+		commonSpec.Output.To.Name = fmt.Sprintf("%s-%s", commonSpec.Output.To.Name, arch)
+
+		build := &buildv1.Build{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-%s", mabc.Name, arch),
+				Namespace: mabc.Namespace,
+				Labels: map[string]string{
+					v1.MultiArchBuildConfigNameLabel: mabc.Name,
+					v1.MultiArchBuildConfigArchLabel: arch,
+				},
+			},
+			Spec: buildv1.BuildSpec{
+				CommonSpec: *commonSpec,
+			},
+		}
+
+		r.logger.WithField("build_namespace", build.Namespace).WithField("build_name", build.Name).Info("Creating build...")
+		if err := r.client.Create(ctx, build); err != nil {
+			return fmt.Errorf("coudldn't create build %s/%s: %w", build.Namespace, build.Name, err)
+		}
+		builds[fmt.Sprintf("%s/%s", build.Namespace, build.Name)] = build
+	}
+
+	r.logger.Info("Watching builds...")
+	w, err := r.buildClient.Builds(mabc.Namespace).Watch(ctx, metav1.ListOptions{
+		LabelSelector: labels.Set{v1.MultiArchBuildConfigNameLabel: mabc.Name}.String(),
+	})
+	if err != nil {
+		return err
+	}
+
+Loop:
+	for {
+		event, ok := <-w.ResultChan()
```

Comment:
@bbguimaraes Do you think that it's better to create another controller that watches/reconciles the Builds instead of running a watcher here?

Reply:
Why not do it in a reconciliation loop?  See also the previous comment.

Reply:
We can add the builds with the multiarchbuildconfig label to be owned by the manager as well. We could potentially trigger the reconciliation process or write another controller for them. That will make things little bit more complicated though. I will dig into it.